### ISSUE TITLE
RFC: Support user defined and extensible deployment strategies

### ISSUE THREAD

**Issue by ironcladlou** [kubernetes/kubernetes#14510]
The new deployment API (#1743) specifies some out of the box strategies (recreate, rolling) which will be implemented inside the deployment controller. It's possible these strategies may not cover all deployment needs for end users, or for downstream users (such as OpenShift). The deployment system should enable:

1. Users to define their own deployment strategy logic in a custom Docker image.
2. Downstream systems to define deployment strategies in a delegate controller.

It's possible these two capabilities are related but orthogonal in their design/implementation. I'm starting the discussion here so we can collect initial thoughts and start linking any existing relevant issues.

**Comment by ironcladlou** [comment]
cc @smarterclayton @nikhiljindal @ghodss @bgrant0607

**Comment by ironcladlou** [comment]
Spoke with @smarterclayton a little more about downstream extensibility, and it seems that as long as the kube deployment controller is scoped to just the strategies supported upstream, adding new downstream strategies could be as simple as just making new controllers which handle other strategy types. If downstream wanted to replace or augment the upstream strategies, the deployment controller could just be deactivated, or reconfigured.

One question I have about that idea as I type this is: how would the user specify a strategy not defined in an upstream constant?

**Comment by smarterclayton** [comment]
Don't strongly validate the strategy.  We need a general pattern for
allowing new fields to be specified that aren't lost on round tripping but
are not strongly versioned by kube.  In the short term, we can allow
arbitrary types and only validate recognized ones, set annotations to
parameterize the types.  In the medium term we should be able to add
extensions.  In the long run we should be able to define extensions and
register them with the api server.

**Comment by bgrant0607** [comment]
I think custom strategies are more relevant to DaemonSet and PetSet (#260).

**Comment by ironcladlou** [comment]
I think it would be enough to provide an interface in the deployment controller for downstream code to plug in custom reconcilers per strategy type. Then kube could accept a deployment using strategy=<anything> and openshift could registry a reconciler to handle it. That would require less validation around the strategy, as @smarterclayton mentioned.

**Comment by smarterclayton** [comment]
A custom strategy is really an evolving triggered job.  The question is -
anything that is totally declarative is a candidate for some sort of
reconcile / adapt path, which of necessity is iterative (one at a time
semantics), with all of the implications of deployment.  So if we don't do
custom strategies in deployment, we're (Openshift) going to need to create
something that can represent it.

On Nov 12, 2015, at 1:11 PM, Dan Mace <notifications@github.com> wrote:

I think it would be enough to provide an interface in the deployment
controller for downstream code to plug in custom reconcilers per strategy
type. Then kube could accept a deployment using strategy= and openshift
could registry a reconciler to handle it. That would require less
validation around the strategy, as @smarterclayton
<https://github.com/smarterclayton> mentioned.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/14510#issuecomment-156187359>
.

**Comment by bgrant0607** [comment]
What kinds of custom strategies do you imagine?

What we've found users want to control, at least for daemons and pets, is timing and sequencing of updates.

**Comment by jimmyjones2** [comment]
@bgrant0607 I've been thinking about Elasticsearch - pods should only be upgraded when all shards are allocated (ie. green status) so that any request can still be serviced. It can take a while for a node to load all of its shards after a restart, although the pod is alive and can service some requests. I presume this use-case is similar to other distributed systems like MongoDB, Kafka etc - pods should only be upgraded when the data is fully replicated.

**Comment by bgrant0607** [comment]
@jimmyjones2 I agree that's an issue. However, stateful apps will be run by a different controller, PetSet:
https://github.com/kubernetes/kubernetes/issues?q=is%3Aopen+is%3Aissue+label%3Aarea%2Fstateful-apps

**Comment by mfojtik** [comment]
@bgrant0607 a custom strategy can be something like using "rolling" as is it today, but add some hooks before, in middle and after the rollout. Inside deployer Pod, we can have logic like this:

```bash
#!/bin/sh

# wait-for-service xyz ?
# backup existing data from PV?
kube-deploy --rollout --until=50%
# now execute something as the rollout is in middle
kube-deploy --rollout --until=100%
# notify-service xyz?
```

**Comment by kargakis** [comment]
What we need for Deployments is auto-pause (assuming it will work at any point in a rollout and not just for canaries) and probably a strategy that delegates to something else.

https://github.com/kubernetes/kubernetes/issues/11505
https://github.com/kubernetes/kubernetes/issues/39337

**Comment by sftim** [comment]
See https://github.com/kubernetes/kubernetes/issues/14512#issuecomment-2540975661 for a potential alternative mechanism to achieve similar outcomes.
